### ISSUE TITLE
Trying to fix Hibernation

### ISSUE THREAD

**Issue by 5T33Z0** [5T33Z0/Lenovo-T530-Hackintosh-OpenCore#48]
@5T33Z0 After spending a lot of time trying to find guides or any info on hibernation in modern hackintosh (OpenCore). I managed to fix hibernation, before these fixes I was only getting a black screen or a restart when trying to wake from hibernation `hibernatemode=25` the main problem was some RTC memory regions getting overwritten when starting the machine again, there were other small issues too. I tested with Catalina and Sonoma and its been reliable so far

To fix hibernation you will the kexts `HibernationFixup` and `RTCMemoryFixup`

- Add `rtcfx_exclude=80-AB` to boot-args
- Set `HibernateMode` to NVRAM in Misc/Boot
- Set `RebuildAppleMemoryMap` to False in Booter/Quirks
- Use the following `ReservedMemory` entry
![Screenshot 2023-12-02 at 8 22 21 PM](https://github.com/5T33Z0/Lenovo-T530-Hackintosh-OpenCore/assets/83997656/d63dd6dc-53cc-4cf2-b971-74975f7e90ec)
[ReservedMemory.plist.zip](https://github.com/5T33Z0/Lenovo-T530-Hackintosh-OpenCore/files/13537889/ReservedMemory.plist.zip)
- Optionally set `HibernateSkipsPicker` in Misc/Boot to true for a more seamless hibernation wake 

To fix a black screen that happens after hibernating and waking once and then try to hibernate again
- Enable the `DiscardHibernateMap` Booter quirk
- Reset NVRAM

_Originally posted by @jozews321 in https://github.com/5T33Z0/Lenovo-T530-Hackintosh-OpenCore/issues/43#issuecomment-1837315676_

**Comment by 5T33Z0** [comment]
Added the settings and kext and changed Hibernatemode to 25 in Sonoma. Now, if I send the system to sleep, I can no longer wake it via the `Fn` key. I guess this is because the keyboard is powered down in Hibernation. If I press the power button, the system comes back right to where I left it. Is this the correct behavior?

Or do I have to change delays o anything else here?:

```
5t33z0@5T33Z0s-MacBook-Pro ~ % pmset -g
System-wide power settings:
Currently in use:
 standbydelaylow      4200
 standby              1
 womp                 1
 halfdim              1
 hibernatefile        /var/vm/sleepimage
 powernap             1
 networkoversleep     0
 disksleep            10
 standbydelayhigh     4200
 sleep                0
 autopoweroffdelay    259200
 hibernatemode        25
 autopoweroff         1
 ttyskeepawake        1
 displaysleep         10
 highstandbythreshold 50
 lidwake              1
```

I don't know how to system to force into hibernating but it damn sure fixed idle sleep.

On the RTCMemoryFixup repo I read: 

> Offsets from 0x80 to 0xAB are used to store some hibernation information (IOHibernateRTCVariables). If any offset in this range causes a conflict, **you can exclude it, but hibernation won't work.

So, if I understand this correctly, with the boot-arg you added, hibernation shouldn't work. 

Originally posted by @5T33Z0  in https://github.com/5T33Z0/Lenovo-T530-Hackintosh-OpenCore/issues/43#issuecomment-1837418515

**Comment by 5T33Z0** [comment]
> @5T33Z0 If I understand correctly with `rtcfx_exclude=80-AB` you are disabling the hibernation using the RTC memory that just causes a lot of trouble
> 
> And with `HibernateMode` set to NVRAM you are only using NVRAM to do the hibernation, that is way more stable than the old RTC stuff.
> 
> Yes I saw that when hibernating the only way to wake the laptop is with the power button, but also the lid sensor wakes it up, so maybe some ACPI fixes can fix that. The delays only apply when using `hibernatemode=3` I think, you can also do some fine tunning with the boot args of `HibernationFixup` to set when and in what conditions the laptop goes to hibernate. And with `hibernatemode=25` it always goes to hibernate Also I forgot to tell that the bios mode needs to be set to UEFI only, CSM disabled for hibernation to work

Originally posted by @jozews321 in https://github.com/5T33Z0/Lenovo-T530-Hackintosh-OpenCore/issues/43#issuecomment-1837486765

**Comment by jozews321** [comment]
@5T33Z0 
No idea how setting HibernateMode to NVRAM fixed idle sleep for you, maybe it was because you had it on Auto, before fixing hibernation I had it on None and idle sleep worked fine

Try to make varios hibernations in a row that's when freezing after wake issues started to happen without the `ReservedMemory` region in my end.

In hibernationmode=3 it's normal that that the laptop darkwakes up when the time in `standbydelayhigh` elapses, in this moment it does the logic if it should enter hibernation or not, the main requirement to make a hibernation in this darkwake is to have the boot-arg `hbfx-ahbm` set to at least 1 if it's not set it's never going into hibernation and the laptop will just sleep again.
I have personally set `hbfx-ahbm` to 5 and both `standbydelay` to 10800 so it wakes up after 3 hours and only enters hibernation when it's disconnected from AC as doing many hibernations can cause excessive wear on the SSD and I mainly use my laptop docked and always connected

I've tried the NVRAM dump but it doesn't work for me either, maybe it needs the Debug version of the kext

As for `autopoweroff` this issue thread goes into detail in some of the things that HibernationFixup can't access https://github.com/acidanthera/bugtracker/issues/2057#issuecomment-1201318837

**Comment by 5T33Z0** [comment]
Yes, I had `Hibernatemode`set to `Auto` before.

`hbfx-ahbm` is set to `1`

If I understand hibernation correctly, it should write the current system state to disk and turn it off after the `autopoweroffdelay` timer has elapsed. So my expectations were:

If I do nothing, the system  enters `sleep` after x seconds, and after x more seconds of `standbydelayhigh` have passed is enters hibernation and after x more seconds for autopoeroff it then turns the system off.

Or am I missing something here in terms of dependencies? I think need to create a flowchart/organigram to fully understand it.

**Comment by 5T33Z0** [comment]
Thanks for the explanations.

I think this flag might be helpful as well:

`DisableStimulusDarkWakeActivityTickle` = `128`: Disable power event kStimulusDarkWakeActivityTickle in kernel, **so this event cannot trigger a switching from dark wake to full wake**

**Comment by flo-at-work** [comment]
Hello and thanks for the work you did on hibernation!

I have a Thinkpad X250 (with Monterey) and a little strange CMOS problem with hibernation, everything works very well, except that when I put it into hibernation (mode 25 or 29) it goes well in hibernation and turn off the power, then when I power on it boots ok (no CMOS error) and resume perfectly from hibernation, BUT then on the next reboot (or shutdown and boot) the BIOS throw me a CMOS Checksum error but pressing Esc it just boots normally.

I managed to find the RTC regions to exclude with RtcRw.efi but what is strange is that it doesn't seems to work, even if I exclude the range 1E-FF (which include the whole RTC but the date) I'm still able to hibernation and I still have the CMOS error on the next restart after resuming from hibernation.

But if I remove the rtcfx_exclude=1E-FF then I even have CMOS error on normal boot/reboot (because offsets 58-59 still needs to be blocked), so it means that rtcfx is indeed blocking offsets, but why am I still able to hibernate and how (and what or where) macOS (or boot.efi) is able to write in the RTC on resuming from hibernation?

I'm on the moment on full debug logs so you can ask me everything.

Thanks in advance!

**Comment by flo-at-work** [comment]
I give here the logs of resuming from hibernation and my config.plist. Maybe someone can have a check if there's anything wrong.

Thanks a lot!

[config.plist.txt](https://github.com/5T33Z0/Lenovo-T530-Hackintosh-OpenCore/files/15094375/config.plist.txt)
[Lilu_1.6.3_21.6.txt](https://github.com/5T33Z0/Lenovo-T530-Hackintosh-OpenCore/files/15094376/Lilu_1.6.3_21.6.txt)
[opencore-2024-04-24-123729.txt](https://github.com/5T33Z0/Lenovo-T530-Hackintosh-OpenCore/files/15094377/opencore-2024-04-24-123729.txt)

**Comment by flo-at-work** [comment]
Hey guys!

So I managed to fix this CMOS checksum issue by modding the BIOS directly, what I did was to simply remove the "DxeCmosInit" module from the UEFI Bios with UEFITool, then flashed back with CH341A programmer. 
So far, no more complains/check about CMOS at Boot (even winning few boot seconds).

Regards

**Comment by 5T33Z0** [comment]
I've had the T530 since 2020 and never had any CMOS issues.

**Comment by flo-at-work** [comment]
Yes sorry if it's a bit off topic, my laptop is a X250...

**Comment by acmehax** [comment]
Very late to the game here but thought that I'd post just in case this helps someone. I have actually used the tips here to fix hibernation on my Dell Precision 5550. Online guides require the Precision 5550 RTC to be unlocked by IFR modification to the system BIOS. However, I was unable to successfully extract any useful IFR information (IFR extraction utilities error out) to locate the bits to modify. This left me with a black screen & KP upon every wake. Following these exact steps:

1. Set HibernateMode to NVRAM in Misc/Boot
2. Set RebuildAppleMemoryMap to False in Booter/Quirks
3. Enable the DiscardHibernateMap Booter quirk

Plus the config.plist changes:

4. NVRAM->Add->4D1FDA02-38C7-4A6A-9CC6-4BCCA8B30102->rtc-blacklist:

80818283 84858687 88898A8B 8C8D8E8F 90919293 94959697 98999A9B 9C9D9E9F A0A1A2A3 A4A5A6A7 A8A9AAAB

7. Kernel->Add: Hibernation.kext
8. Kernel->Add: RTCMemoryFixup.kext

and these are my pmsets:

System-wide power settings:
Currently in use:
 standbydelaylow      1
 standby              1
 acwake               0
 halfdim              1
 hibernatefile        /var/vm/sleepimage
 proximitywake        0
 powernap             1
 disksleep            10
 standbydelayhigh     1
 sleep                1 (sleep prevented by sharingd)
 hibernatemode        25
 ttyskeepawake        1
 displaysleep         3
 tcpkeepalive         1
 highstandbythreshold 50
 lowpowermode         0
 lidwake              1

----------------

Now, while this all worked beautifully under Moneterey, it fails miserably for Ventura.
So, the research continues ...

**Comment by 5T33Z0** [comment]
@acmehax Thanks for your input. I am just not sure if these settings are applicable to the T530, since the Dell precision 5500 uses a 10th Gen Intel CPU, while mine uses a 3rd Gen. Did you mean `HibernationFixup.kext` when you wrote `Hibernation.kext`, or is it a different ext?

**Comment by acmehax** [comment]
Yes, you are correct, that should have read HibernationFixup.kext (fixed now). Yes, I understand that this has nothing to do with the Lenovo and I don't wish to hijack this topic but I find that I often happen upon tips when perusing threads for other systems so I figured prhaps this helps more than hurts for now. I am a bit confused now as it seems that blacklisting thje memory range that I did should have killed hiberanation altogether (as per RTCMemoryFixup.kext details) but under Monterey, I swear hiberantion was functional, requiring a power button click to restore desktop. Now, I've removed the blacklist and tried the reserved memory mod but I still have black screen wake and likely KP as fan runs full on wake.

**Comment by zx0r** [comment]
> [@acmehax](https://github.com/acmehax) Thanks for your input. I am just not sure if these settings are applicable to the T530, since the Dell precision 5500 uses a 10th Gen Intel CPU, while mine uses a 3rd Gen. Did you mean `HibernationFixup.kext` when you wrote `Hibernation.kext`, or is it a different ext?

[Hackintosh Power Management](https://github.com/zx0r/HackintoshBible/blob/main/PowerManagement/README.md)
Hi bro
Your repository has been very helpful in fine tuning my Hackintosh. 
Check out my repo, maybe this information will be useful.
In my case,  sleep and hibernation works perfectly fine in all modes without waking up.

**Comment by 5T33Z0** [comment]
@zx0r Thanks. I added a link to ot in my OC-Little repo (Chapt. 4)

**Comment by 5T33Z0** [comment]
@jozews321 Hey, sorry to bother you this way. But could you please post a compiled version of your fork of YogaSMC (with prefpane, helper and app) for me here, please? I tried to compile it myself but it fails and I don'tknowhow to fix it.

Edit: Nevermind, I managed to build it.

**Comment by jozews321** [comment]
@5T33Z0 Hi, sorry for the late reply, have you found any issues in the new DYTC functions in the menubar app?. 

I'm currently trying to rewrite parts of the prefpane that are broken in newer OSs (at least for me in the T480), like the DYTC slider that doesn't apply the new power setting, and the setting to change the battery charge thresholds that just resets when I set a value and doesn't get applied. 

Have you experienced those issues as well?

**Comment by 5T33Z0** [comment]
@jozews321 Yes. For me, the app and the prepanes crash. The App crashes as soon as I move any slider. And if I re-open the app, no sliders are available. But this could be co-incidental because the prefpane is the actual thing that is not working/crashing.

**Comment by jozews321** [comment]
> [@jozews321](https://github.com/jozews321) Yes. For me, the app and the prepanes crash. The App crashes as soon as I move any slider. And if I re-open the app, no sliders are available. But this could be co-incidental because the prefpane is the actual thing that is not working/crashing.

Yes, the prefpane is broken in newer macOS versions, I'm currently trying to fix it and better integrate it with the new changes in the menubar app
